RABBITHOLE.md

```diff
@@ -0,0 +1,93 @@
+* The service-based listener use case obviously needs a container, and PSR-11 takes care of everything we need.  Which... suggests PSR-11 may have been a good thing after all. :-)
+* The compiled use-case is handled best not by a compiled dispatcher but a compiled ListenerSet.
+* In fact, with listener sets (aggregators, whatever) factored out to a separate interface a nearly universal dispatcher is stupid-easy to write.  The only use case for an alternate implementation seems to be for instrumentation (eg, logging every event directly for debugging).  I consider this a good thing.
+* However, there seems little value to a unified registration interface.  Every use case (unordered, priority ordered, before/after ordered, services, etc.) seems to require a different method signature.  That suggests we should do one of the following:
+** Do not provide a registration interface at all, and explicitly kick that downstream to implementers.  This has the benefit of flexibility but the downside of having no standardized way for trivial cases to register themselves.
+** Provide multiple "common case" registration interfaces (probably unordered and priority ordered) and then explicitly say that they're optional and it's totally legit to define your own instead.
+* The word "listener" is far too easy to misspell as "listner".
+* The code in pretty much all implementations is scarily simple.  That is a good sign.
+
+## The interfaces
+
+```php
+interface EventInterface
+{
+    public function stopPropagation(bool $stop = true) : self;
+
+    public function stopped() : bool;
+}
+```
+
+A trivially simple interface with propagation control.  There's also a trait provided that implements them both in the only logical way, which is a natural fit for a util library.
+
+```php
+interface ListenerSetInterface
+{
+    public function getListenersFor(EventInterface $event): iterable;
+}
+```
+
+This is a read-only collection of listeners.  It's basically the same as Matthew's "aggregate", as that turned out to be a really elegant solution but I find "aggregate" to be too pretentious a word even for me :-).  The only thing you can do is get a list of listeners that are applicable to an event object.  It returns an iterable so that implementers can use an array or generator or whatever at their leisure.
+
+```php
+interface DispatcherInterface
+{
+    public function dispatch(EventInterface $event) : EventInterface;
+}
+```
+
+The dispatcher interface.  Give it an event, it runs it through the appropriate listeners and then gives the event back.  I have it returning the event to make chaining easier.  Eg:
+
+```php
+$event = new Whatever();
+$results = $dispatcher->dispatch($event)->getResults();
+```
+
+I am not wedded to that but it seems like a nice convenience for the "give me your stuff" and "modify my thingie" use cases.
+
+```php
+interface BasicRegistrationInterface
+{
+    public function addListener(callable $listener, string $type = null) : void;
```

Comment:
I know we have `$type` second so that we can make it optional. However, this signature is now backwards compared to most implementations (in most langs) - e.g.

```js
$('#foo').on('click', function() {});
```

That troubles me. And, selfishly, we will not use the "auto-detect" in Symfony. Well, we will have this, but the auto-detection is elsewhere, before loading up the dispatcher. So, it's not needed for us.